packages/legend-query-builder/src/stores/fetch-structure/tds/projection/QueryBuilderProjectionStateBuilder.ts

```diff
@@ -96,47 +86,132 @@ export const processTDSProjectExpression = (
+  queryBuilderState: QueryBuilderState,
+  parentLambda: LambdaFunction,
+): void => {
+  // update fetch-structure
+  queryBuilderState.fetchStructureState.changeImplementation(
+    FETCH_STRUCTURE_IMPLEMENTATION.TABULAR_DATA_STRUCTURE,
+  );
+  // check parameters
+  assertTrue(
+    expression.parametersValues.length === 2,
+    `Can't process project() expression: project() expects 2 arguments`,
+  );
+
+  validateTDSProjectPrecedingExpression(
+    expression,
+    queryBuilderState,
+    parentLambda,
+  );
 
   // check columns
   const columnLambdas = expression.parametersValues[1];
-  assertType(
-    columnLambdas,
-    CollectionInstanceValue,
-    `Can't process project() expression: project() expects argument #1 to be a collection`,
-  );
-  columnLambdas.values.map((value) =>
+  if (columnLambdas instanceof CollectionInstanceValue) {
+    columnLambdas.values.map((value) =>
+      QueryBuilderValueSpecificationProcessor.processChild(
+        value,
+        expression,
+        parentLambda,
+        queryBuilderState,
+      ),
+    );
+  } else {
+    assertType(
+      columnLambdas,
+      SimpleFunctionExpression,
+      `Can't process project() expression: project() expects argument #1 to be a function expression`,
+    );
     QueryBuilderValueSpecificationProcessor.processChild(
-      value,
+      columnLambdas,
       expression,
       parentLambda,
       queryBuilderState,
-    ),
-  );
+    );
+  }
+};
 
-  // check column aliases
-  const columnAliases = expression.parametersValues[2];
-  assertType(
-    columnAliases,
-    CollectionInstanceValue,
-    `Can't process project() expression: project() expects argument #2 to be a collection`,
-  );
-  assertTrue(
-    columnLambdas.values.length === columnAliases.values.length,
-    `Can't process project() expression: number of aliases does not match the number of columns`,
-  );
-  const aliases = columnAliases.values
-    .map(extractNullableStringFromInstanceValue)
-    .filter(isNonNullable);
+export const processTDSProjectExpression = (
+  expression: SimpleFunctionExpression,
+  queryBuilderState: QueryBuilderState,
+  parentLambda: LambdaFunction,
+): void => {
+  if (expression.parametersValues.length === 2) {
+    processTDSProjectColExpression(expression, queryBuilderState, parentLambda);
+  } else {
+    // update fetch-structure
+    queryBuilderState.fetchStructureState.changeImplementation(
+      FETCH_STRUCTURE_IMPLEMENTATION.TABULAR_DATA_STRUCTURE,
+    );
 
-  // build state
-  if (
-    queryBuilderState.fetchStructureState.implementation instanceof
-    QueryBuilderTDSState
-  ) {
-    const tdsState = queryBuilderState.fetchStructureState.implementation;
-    tdsState.projectionColumns.forEach((column, idx) =>
-      column.setColumnName(aliases[idx] as string),
+    // check parameters
+    assertTrue(
+      expression.parametersValues.length === 3,
+      `Can't process project() expression: project() expects 2 arguments`,
```

Comment:
Why are we checking that `expression.parametersValues.length === 3` but then in the error message we say it expects 2 arguments?